{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6de7a307",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>SMILES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cannflavin B</td>\n",
       "      <td>CC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C(C=C3)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cannflavin C</td>\n",
       "      <td>CC(=CCCC(=CCC1=C2C(=C(C=C1O)O)C(=O)C=C(O2)C3=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Canniprene</td>\n",
       "      <td>CC(=CCC1=C(C=CC(=C1O)OC)CCC2=CC(=CC(=C2)OC)O)C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cannabistilbene I</td>\n",
       "      <td>CC(=CCC1=C(C=CC(=C1)CCC2=CC(=CC(=C2)OC)O)O)C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6-prenylnaringenin</td>\n",
       "      <td>CC(=CCC1=C(C2=C(C=C1O)OC(CC2=O)C3=CC=C(C=C3)O)O)C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hesperetin</td>\n",
       "      <td>COC1=C(C=C(C=C1)C2CC(=O)C3=C(C=C(C=C3O2)O)O)O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Luteolin</td>\n",
       "      <td>C1=CC(=C(C=C1C2=CC(=O)C3=C(C=C(C=C3O2)O)O)O)O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Quercetin</td>\n",
       "      <td>C1=CC(=C(C=C1C2=C(C(=O)C3=C(C=C(C=C3O2)O)O)O)O)O</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Name                                             SMILES\n",
       "0        Cannflavin A  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...\n",
       "1        Cannflavin B  CC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C(C=C3)...\n",
       "2        Cannflavin C  CC(=CCCC(=CCC1=C2C(=C(C=C1O)O)C(=O)C=C(O2)C3=C...\n",
       "3          Canniprene     CC(=CCC1=C(C=CC(=C1O)OC)CCC2=CC(=CC(=C2)OC)O)C\n",
       "4   Cannabistilbene I       CC(=CCC1=C(C=CC(=C1)CCC2=CC(=CC(=C2)OC)O)O)C\n",
       "5  6-prenylnaringenin  CC(=CCC1=C(C2=C(C=C1O)OC(CC2=O)C3=CC=C(C=C3)O)O)C\n",
       "6          Hesperetin      COC1=C(C=C(C=C1)C2CC(=O)C3=C(C=C(C=C3O2)O)O)O\n",
       "7            Luteolin      C1=CC(=C(C=C1C2=CC(=O)C3=C(C=C(C=C3O2)O)O)O)O\n",
       "8           Quercetin   C1=CC(=C(C=C1C2=C(C(=O)C3=C(C=C(C=C3O2)O)O)O)O)O"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "fasta = open(\"disease.fasta\")\n",
    "df_drugs = pd.read_csv('cann_mols.csv')  \n",
    "\n",
    "fasta_list = list(SimpleFastaParser(fasta))\n",
    "\n",
    "protein_df = pd.DataFrame(fasta_list, columns=['info', 'sequence'])\n",
    "\n",
    "def string_splitter(string):\n",
    "    string = string.split(\"HUMAN\",1)[1]\n",
    "    string = string.split(\"OS=Homo sapiens\")\n",
    "    \n",
    "    return string\n",
    "\n",
    "def info_parser(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info'] # rename the three columns # rename the three columns\n",
    "\n",
    "    return df\n",
    "\n",
    "def info_parser_200k(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info', 'none'] # rename the three columns # rename the three columns\n",
    "    df.drop(\"none\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    return df\n",
    "\n",
    "def info_pre_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of df\n",
    "    list_of_info = list(dfx['info']) # convert column values to list\n",
    "    list_of_info = [string_splitter(x) for x in list_of_info] # apply string_splitter function to list elements\n",
    "    df['temp_col'] = list_of_info # create a temporary column from the processed list\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    split_df = pd.DataFrame(df['temp_col'].tolist(), columns=['Protein name', 'info']) # creating this dataframe to merge back onto processed dataframe\n",
    "    df.drop(\"temp_col\", axis=1, inplace=True) # dropping temp_col\n",
    "    df = pd.concat([df, split_df], axis=1) # merging both dataframes\n",
    "\n",
    "    return df\n",
    "    \n",
    "def info_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.columns = ['type', 'id', 'Protein name', 'drop', 'Species', 'Gene', 'PE', 'Mutation'] # rename columns\n",
    "    df.drop(\"drop\", axis=1, inplace=True) # drop empty column\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.drop('Species', axis=1, inplace=True) # drop column\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "def info_processed_200k(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df = df.drop(df.columns[[0, 3, 4, 6, 7, 8, 9, 10, 11, 12]],axis = 1)\n",
    "\n",
    "    df.columns = ['id', 'Protein name', 'Gene'] # rename columns\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "def info_processed_disease(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df = df.drop(df.columns[[0, 3, 4, 6, 7, 8, 9]],axis = 1)\n",
    "\n",
    "    df.columns = ['id', 'Protein name', 'Gene'] # rename columns\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "#clean_proteins = info_parser_200k(protein_df)\n",
    "                 #pipe(info_parser).\n",
    "                 #pipe(info_pre_processed).\n",
    "                 #pipe(info_processed))\n",
    "            \n",
    "clean_proteins = (protein_df.\n",
    "                 pipe(info_parser).\n",
    "                 pipe(info_pre_processed).\n",
    "                 pipe(info_processed_disease))\n",
    "\n",
    "\n",
    "\n",
    "df_targets = clean_proteins[['Protein name','sequence']].copy() # making a copy of the clean dataframe.\n",
    "\n",
    "target_name = df_targets['Protein name'].tolist()\n",
    "target = df_targets.sequence.tolist()\n",
    "drug_name = df_drugs.Name.tolist()\n",
    "drug = df_drugs.SMILES.tolist()\n",
    "\n",
    "df_targets.to_csv('disease_targets.csv')\n",
    "clean_proteins.to_csv('clean_proteins_disease.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ee4afa61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A024QZB8</td>\n",
       "      <td>Battenin</td>\n",
       "      <td>CLN3</td>\n",
       "      <td>MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A218KGR2</td>\n",
       "      <td>Amyloid-beta A4 protein</td>\n",
       "      <td>APP</td>\n",
       "      <td>MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A5B9</td>\n",
       "      <td>T cell receptor beta constant 2</td>\n",
       "      <td>TRBC2</td>\n",
       "      <td>DLKNVFPPKVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0AV02</td>\n",
       "      <td>Solute carrier family 12 member 8</td>\n",
       "      <td>SLC12A8</td>\n",
       "      <td>MTQMSQVQELFHEAAQQDALAQPQPWWKTQLFMWEPVLFGTWDGVF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0PJY2</td>\n",
       "      <td>Fez family zinc finger protein 1</td>\n",
       "      <td>FEZF1</td>\n",
       "      <td>MDSSCHNATTKMLATAPARGNMMSTSKPLAFSIERIMARTPEPKAL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6631</th>\n",
       "      <td>X5CF33</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>MVDFPGYNLSGAVASFLFILLTMKQSEDFRVIGPAHPILAGVGEDA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6632</th>\n",
       "      <td>X5CF57</td>\n",
       "      <td>BRD2</td>\n",
       "      <td>BRD2</td>\n",
       "      <td>MASVPALQLTPANPPPPEVSNPKKPGRVTNQLQYLHKVVMKALWKH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6633</th>\n",
       "      <td>X5CKA9</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>MMVLQVSAAPRTVALTALLMVLLTSVVQGRATPENYVHQLRQECYA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6634</th>\n",
       "      <td>X5D146</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>MVDFPGYNLSGAVASFLFILLTMKQSEDFRVIGPAHPILAGVGEDA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6635</th>\n",
       "      <td>X5D2T4</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>MMVLQVSAAPRTVALTALLMVLLTSVVQGRATP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6636 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                         Protein name      Gene  \\\n",
       "0     A0A024QZB8                            Battenin       CLN3   \n",
       "1     A0A218KGR2             Amyloid-beta A4 protein        APP   \n",
       "2         A0A5B9     T cell receptor beta constant 2      TRBC2   \n",
       "3         A0AV02   Solute carrier family 12 member 8    SLC12A8   \n",
       "4         A0PJY2    Fez family zinc finger protein 1      FEZF1   \n",
       "...          ...                                  ...       ...   \n",
       "6631      X5CF33                               BTNL2      BTNL2   \n",
       "6632      X5CF57                                BRD2       BRD2   \n",
       "6633      X5CKA9     MHC class II antigen (Fragment)   HLA-DPB1   \n",
       "6634      X5D146                               BTNL2      BTNL2   \n",
       "6635      X5D2T4     MHC class II antigen (Fragment)   HLA-DPB1   \n",
       "\n",
       "                                               sequence  \n",
       "0     MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...  \n",
       "1     MLPGLALLLLAAWTARALEVPTDGNAGLLAEPQIAMFCGRLNMHMN...  \n",
       "2     DLKNVFPPKVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWV...  \n",
       "3     MTQMSQVQELFHEAAQQDALAQPQPWWKTQLFMWEPVLFGTWDGVF...  \n",
       "4     MDSSCHNATTKMLATAPARGNMMSTSKPLAFSIERIMARTPEPKAL...  \n",
       "...                                                 ...  \n",
       "6631  MVDFPGYNLSGAVASFLFILLTMKQSEDFRVIGPAHPILAGVGEDA...  \n",
       "6632  MASVPALQLTPANPPPPEVSNPKKPGRVTNQLQYLHKVVMKALWKH...  \n",
       "6633  MMVLQVSAAPRTVALTALLMVLLTSVVQGRATPENYVHQLRQECYA...  \n",
       "6634  MVDFPGYNLSGAVASFLFILLTMKQSEDFRVIGPAHPILAGVGEDA...  \n",
       "6635                  MMVLQVSAAPRTVALTALLMVLLTSVVQGRATP  \n",
       "\n",
       "[6636 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def info_processed_disease(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df = df.drop(df.columns[[0, 3, 4, 6, 7, 8, 9]],axis = 1)\n",
    "\n",
    "    df.columns = ['id', 'Protein name', 'Gene'] # rename columns\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "#clean_proteins = info_parser_200k(protein_df)\n",
    "                 #pipe(info_parser).\n",
    "                 #pipe(info_pre_processed).\n",
    "                 #pipe(info_processed))\n",
    "            \n",
    "clean_proteins = (protein_df.\n",
    "                 pipe(info_parser).\n",
    "                 pipe(info_pre_processed))\n",
    "\n",
    "test = info_processed_200k(clean_proteins)\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24735593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr</td>\n",
       "      <td>A0A024QZB8</td>\n",
       "      <td>Battenin</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=CLN3</td>\n",
       "      <td>PE=1</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr</td>\n",
       "      <td>A0A218KGR2</td>\n",
       "      <td>Amyloid-beta A4 protein</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=APP</td>\n",
       "      <td>PE=3</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0A5B9</td>\n",
       "      <td>T cell receptor beta constant 2</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=TRBC2</td>\n",
       "      <td>PE=1</td>\n",
       "      <td>SV=2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0AV02</td>\n",
       "      <td>Solute carrier family 12 member 8</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=SLC12A8</td>\n",
       "      <td>PE=1</td>\n",
       "      <td>SV=4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0PJY2</td>\n",
       "      <td>Fez family zinc finger protein 1</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=FEZF1</td>\n",
       "      <td>PE=1</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6631</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CF33</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=BTNL2</td>\n",
       "      <td>PE=4</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6632</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CF57</td>\n",
       "      <td>BRD2</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=BRD2</td>\n",
       "      <td>PE=4</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6633</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CKA9</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=HLA-DPB1</td>\n",
       "      <td>PE=4</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6634</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5D146</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=BTNL2</td>\n",
       "      <td>PE=4</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6635</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5D2T4</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td></td>\n",
       "      <td>OX=9606</td>\n",
       "      <td>GN=HLA-DPB1</td>\n",
       "      <td>PE=4</td>\n",
       "      <td>SV=1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6636 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     type          id                         Protein name 0         1  \\\n",
       "0      tr  A0A024QZB8                            Battenin      OX=9606   \n",
       "1      tr  A0A218KGR2             Amyloid-beta A4 protein      OX=9606   \n",
       "2      sp      A0A5B9     T cell receptor beta constant 2      OX=9606   \n",
       "3      sp      A0AV02   Solute carrier family 12 member 8      OX=9606   \n",
       "4      sp      A0PJY2    Fez family zinc finger protein 1      OX=9606   \n",
       "...   ...         ...                                  ... ..      ...   \n",
       "6631   tr      X5CF33                               BTNL2      OX=9606   \n",
       "6632   tr      X5CF57                                BRD2      OX=9606   \n",
       "6633   tr      X5CKA9     MHC class II antigen (Fragment)      OX=9606   \n",
       "6634   tr      X5D146                               BTNL2      OX=9606   \n",
       "6635   tr      X5D2T4     MHC class II antigen (Fragment)      OX=9606   \n",
       "\n",
       "                2     3     4     5     6  \n",
       "0         GN=CLN3  PE=1  SV=1  None  None  \n",
       "1          GN=APP  PE=3  SV=1  None  None  \n",
       "2        GN=TRBC2  PE=1  SV=2  None  None  \n",
       "3      GN=SLC12A8  PE=1  SV=4  None  None  \n",
       "4        GN=FEZF1  PE=1  SV=1  None  None  \n",
       "...           ...   ...   ...   ...   ...  \n",
       "6631     GN=BTNL2  PE=4  SV=1  None  None  \n",
       "6632      GN=BRD2  PE=4  SV=1  None  None  \n",
       "6633  GN=HLA-DPB1  PE=4  SV=1  None  None  \n",
       "6634     GN=BTNL2  PE=4  SV=1  None  None  \n",
       "6635  GN=HLA-DPB1  PE=4  SV=1  None  None  \n",
       "\n",
       "[6636 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5133b01a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tr</td>\n",
       "      <td>A0A024QZB8</td>\n",
       "      <td>Battenin</td>\n",
       "      <td>OX=9606 GN=CLN3 PE=1 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tr</td>\n",
       "      <td>A0A218KGR2</td>\n",
       "      <td>Amyloid-beta A4 protein</td>\n",
       "      <td>OX=9606 GN=APP PE=3 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0A5B9</td>\n",
       "      <td>T cell receptor beta constant 2</td>\n",
       "      <td>OX=9606 GN=TRBC2 PE=1 SV=2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0AV02</td>\n",
       "      <td>Solute carrier family 12 member 8</td>\n",
       "      <td>OX=9606 GN=SLC12A8 PE=1 SV=4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp</td>\n",
       "      <td>A0PJY2</td>\n",
       "      <td>Fez family zinc finger protein 1</td>\n",
       "      <td>OX=9606 GN=FEZF1 PE=1 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6631</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CF33</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>OX=9606 GN=BTNL2 PE=4 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6632</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CF57</td>\n",
       "      <td>BRD2</td>\n",
       "      <td>OX=9606 GN=BRD2 PE=4 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6633</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5CKA9</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td>OX=9606 GN=HLA-DPB1 PE=4 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6634</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5D146</td>\n",
       "      <td>BTNL2</td>\n",
       "      <td>OX=9606 GN=BTNL2 PE=4 SV=1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6635</th>\n",
       "      <td>tr</td>\n",
       "      <td>X5D2T4</td>\n",
       "      <td>MHC class II antigen (Fragment)</td>\n",
       "      <td>OX=9606 GN=HLA-DPB1 PE=4 SV=1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6636 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     type          id                         Protein name  \\\n",
       "0      tr  A0A024QZB8                            Battenin    \n",
       "1      tr  A0A218KGR2             Amyloid-beta A4 protein    \n",
       "2      sp      A0A5B9     T cell receptor beta constant 2    \n",
       "3      sp      A0AV02   Solute carrier family 12 member 8    \n",
       "4      sp      A0PJY2    Fez family zinc finger protein 1    \n",
       "...   ...         ...                                  ...   \n",
       "6631   tr      X5CF33                               BTNL2    \n",
       "6632   tr      X5CF57                                BRD2    \n",
       "6633   tr      X5CKA9     MHC class II antigen (Fragment)    \n",
       "6634   tr      X5D146                               BTNL2    \n",
       "6635   tr      X5D2T4     MHC class II antigen (Fragment)    \n",
       "\n",
       "                                info  \n",
       "0          OX=9606 GN=CLN3 PE=1 SV=1  \n",
       "1           OX=9606 GN=APP PE=3 SV=1  \n",
       "2         OX=9606 GN=TRBC2 PE=1 SV=2  \n",
       "3       OX=9606 GN=SLC12A8 PE=1 SV=4  \n",
       "4         OX=9606 GN=FEZF1 PE=1 SV=1  \n",
       "...                              ...  \n",
       "6631      OX=9606 GN=BTNL2 PE=4 SV=1  \n",
       "6632       OX=9606 GN=BRD2 PE=4 SV=1  \n",
       "6633   OX=9606 GN=HLA-DPB1 PE=4 SV=1  \n",
       "6634      OX=9606 GN=BTNL2 PE=4 SV=1  \n",
       "6635   OX=9606 GN=HLA-DPB1 PE=4 SV=1  \n",
       "\n",
       "[6636 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62dc3af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_targets['drug_name'] = df_drugs['Name'][0] # adding a column with a constant value of \"cannaflavin a\".\n",
    "df_targets['SMILES'] = df_drugs['SMILES'][0] # adding a column with a constant value of \"cannaflavin a\" SMILES string.\n",
    "\n",
    "df_targets_2 = df_targets.copy()\n",
    "df_targets_2['drug_name'] = df_drugs['Name'][1]\n",
    "df_targets_2['SMILES'] = df_drugs['SMILES'][1]\n",
    "\n",
    "df_targets_3 = df_targets.copy()\n",
    "df_targets_3['drug_name'] = df_drugs['Name'][2]\n",
    "df_targets_3['SMILES'] = df_drugs['SMILES'][2]\n",
    "\n",
    "df_targets_4 = df_targets.copy()\n",
    "df_targets_4['drug_name'] = df_drugs['Name'][3]\n",
    "df_targets_4['SMILES'] = df_drugs['SMILES'][3]\n",
    "\n",
    "df_targets_5 = df_targets.copy()\n",
    "df_targets_5['drug_name'] = df_drugs['Name'][4]\n",
    "df_targets_5['SMILES'] = df_drugs['SMILES'][4]\n",
    "\n",
    "df_targets_6 = df_targets.copy()\n",
    "df_targets_6['drug_name'] = df_drugs['Name'][5]\n",
    "df_targets_6['SMILES'] = df_drugs['SMILES'][5]\n",
    "\n",
    "df_targets_7 = df_targets.copy()\n",
    "df_targets_7['drug_name'] = df_drugs['Name'][6]\n",
    "df_targets_7['SMILES'] = df_drugs['SMILES'][6]\n",
    "\n",
    "df_targets_8 = df_targets.copy()\n",
    "df_targets_8['drug_name'] = df_drugs['Name'][7]\n",
    "df_targets_8['SMILES'] = df_drugs['SMILES'][7]\n",
    "\n",
    "df_targets_9 = df_targets.copy()\n",
    "df_targets_9['drug_name'] = df_drugs['Name'][8]\n",
    "df_targets_9['SMILES'] = df_drugs['SMILES'][8]\n",
    "\n",
    "targets = pd.concat([df_targets, \n",
    "                  df_targets_2, \n",
    "                  df_targets_3, \n",
    "                  df_targets_4, \n",
    "                  df_targets_5, \n",
    "                  df_targets_6, \n",
    "                  df_targets_7, \n",
    "                  df_targets_8, \n",
    "                  df_targets_9], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "967fd792",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1844154"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(target_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7c98afcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein name</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T-box 2, isoform CRA_a</td>\n",
       "      <td>MREPALAASAMAYHPFHAPRPADFPMSAFLAAAQPSFFPALALPPG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Receptor protein-tyrosine kinase</td>\n",
       "      <td>MELQAARACFALLWGCALAAAAAAQGKEVVLLDFAAAGGELGWLTH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Battenin</td>\n",
       "      <td>MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sirtuin (Silent mating type information regul...</td>\n",
       "      <td>MADEAALALQPGGSPSAAGADREAASSPAGEPLRKRPRRDGPGLER...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Phospholipid phosphatase 1</td>\n",
       "      <td>MFDKTRLPYVALDVLCVLLAGLPFAILTSRHTPFQRGVFCNDESIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204901</th>\n",
       "      <td>Probable ATP-dependent RNA helicase DDX5</td>\n",
       "      <td>MSGYSSDRDRGRDRGFGAPRFGGSRAGPLSGKKFGNPGEKLVKKKW...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204902</th>\n",
       "      <td>ELKS/Rab6-interacting/CAST family member 1</td>\n",
       "      <td>MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204903</th>\n",
       "      <td>Voltage-dependent calcium channel subunit alp...</td>\n",
       "      <td>MKLEFLQRKFWAATRQCSTVDGPCTQSCEDSDLDCFVIDNNGFILI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204904</th>\n",
       "      <td>ELKS/Rab6-interacting/CAST family member 1</td>\n",
       "      <td>MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204905</th>\n",
       "      <td>Mitochondrial enolase superfamily member 1 (F...</td>\n",
       "      <td>XHGADAMHTDPDYSAAYVVIETDAEDGIKGCGITFTLGKGTEVDWS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204906 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Protein name  \\\n",
       "0                                 T-box 2, isoform CRA_a    \n",
       "1                       Receptor protein-tyrosine kinase    \n",
       "2                                               Battenin    \n",
       "3        Sirtuin (Silent mating type information regul...   \n",
       "4                             Phospholipid phosphatase 1    \n",
       "...                                                   ...   \n",
       "204901          Probable ATP-dependent RNA helicase DDX5    \n",
       "204902        ELKS/Rab6-interacting/CAST family member 1    \n",
       "204903   Voltage-dependent calcium channel subunit alp...   \n",
       "204904        ELKS/Rab6-interacting/CAST family member 1    \n",
       "204905   Mitochondrial enolase superfamily member 1 (F...   \n",
       "\n",
       "                                                 sequence  \n",
       "0       MREPALAASAMAYHPFHAPRPADFPMSAFLAAAQPSFFPALALPPG...  \n",
       "1       MELQAARACFALLWGCALAAAAAAQGKEVVLLDFAAAGGELGWLTH...  \n",
       "2       MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...  \n",
       "3       MADEAALALQPGGSPSAAGADREAASSPAGEPLRKRPRRDGPGLER...  \n",
       "4       MFDKTRLPYVALDVLCVLLAGLPFAILTSRHTPFQRGVFCNDESIK...  \n",
       "...                                                   ...  \n",
       "204901  MSGYSSDRDRGRDRGFGAPRFGGSRAGPLSGKKFGNPGEKLVKKKW...  \n",
       "204902  MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...  \n",
       "204903  MKLEFLQRKFWAATRQCSTVDGPCTQSCEDSDLDCFVIDNNGFILI...  \n",
       "204904  MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...  \n",
       "204905  XHGADAMHTDPDYSAAYVVIETDAEDGIKGCGITFTLGKGTEVDWS...  \n",
       "\n",
       "[204906 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_proteins\n",
    "df_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e270d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "fasta = open(\"uniprot_200k.fasta\")\n",
    "df_drugs = pd.read_csv('cann_mols.csv')  \n",
    "\n",
    "fasta_list = list(SimpleFastaParser(fasta))\n",
    "\n",
    "protein_df = pd.DataFrame(fasta_list, columns=['info', 'sequence'])\n",
    "\n",
    "def string_splitter(string):\n",
    "    string = string.split(\"HUMAN\",1)[1]\n",
    "    string = string.split(\"OS=Homo sapiens\")\n",
    "    \n",
    "    return string\n",
    "\n",
    "def info_parser(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info'] # rename the three columns # rename the three columns\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def info_parser_200k(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info', 'none'] # rename the three columns # rename the three columns\n",
    "    df.drop(\"none\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    return df\n",
    "\n",
    "def info_pre_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of df\n",
    "    list_of_info = list(dfx['info']) # convert column values to list\n",
    "    list_of_info = [string_splitter(x) for x in list_of_info] # apply string_splitter function to list elements\n",
    "    df['temp_col'] = list_of_info # create a temporary column from the processed list\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    split_df = pd.DataFrame(df['temp_col'].tolist(), columns=['Protein name', 'info']) # creating this dataframe to merge back onto processed dataframe\n",
    "    df.drop(\"temp_col\", axis=1, inplace=True) # dropping temp_col\n",
    "    df = pd.concat([df, split_df], axis=1) # merging both dataframes\n",
    "\n",
    "    return df\n",
    "    \n",
    "def info_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.columns = ['type', 'id', 'Protein name', 'drop', 'Species', 'Gene', 'PE', 'Mutation'] # rename columns\n",
    "    df.drop(\"drop\", axis=1, inplace=True) # drop empty column\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.drop('Species', axis=1, inplace=True) # drop column\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "def info_processed_200k(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df = df.drop(df.columns[[0, 3, 4, 6, 7, 8, 9, 10, 11, 12]],axis = 1)\n",
    "\n",
    "    df.columns = ['id', 'Protein name', 'Gene'] # rename columns\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "#clean_proteins = info_parser_200k(protein_df)\n",
    "                 #pipe(info_parser).\n",
    "                 #pipe(info_pre_processed).\n",
    "                 #pipe(info_processed))\n",
    "            \n",
    "clean_proteins = (protein_df.\n",
    "                 pipe(info_parser_200k).\n",
    "                 pipe(info_pre_processed).\n",
    "                 pipe(info_processed_200k))\n",
    "\n",
    "df_targets = clean_proteins[['Protein name','sequence']].copy() # making a copy of the clean dataframe.\n",
    "\n",
    "df_targets['drug_name'] = df_drugs['Name'][0] # adding a column with a constant value of \"cannaflavin a\".\n",
    "df_targets['SMILES'] = df_drugs['SMILES'][0] # adding a column with a constant value of \"cannaflavin a\" SMILES string.\n",
    "\n",
    "df_targets_2 = df_targets.copy()\n",
    "df_targets_2['drug_name'] = df_drugs['Name'][1]\n",
    "df_targets_2['SMILES'] = df_drugs['SMILES'][1]\n",
    "\n",
    "df_targets_3 = df_targets.copy()\n",
    "df_targets_3['drug_name'] = df_drugs['Name'][2]\n",
    "df_targets_3['SMILES'] = df_drugs['SMILES'][2]\n",
    "\n",
    "df_targets_4 = df_targets.copy()\n",
    "df_targets_4['drug_name'] = df_drugs['Name'][3]\n",
    "df_targets_4['SMILES'] = df_drugs['SMILES'][3]\n",
    "\n",
    "df_targets_5 = df_targets.copy()\n",
    "df_targets_5['drug_name'] = df_drugs['Name'][4]\n",
    "df_targets_5['SMILES'] = df_drugs['SMILES'][4]\n",
    "\n",
    "df_targets_6 = df_targets.copy()\n",
    "df_targets_6['drug_name'] = df_drugs['Name'][5]\n",
    "df_targets_6['SMILES'] = df_drugs['SMILES'][5]\n",
    "\n",
    "df_targets_7 = df_targets.copy()\n",
    "df_targets_7['drug_name'] = df_drugs['Name'][6]\n",
    "df_targets_7['SMILES'] = df_drugs['SMILES'][6]\n",
    "\n",
    "df_targets_8 = df_targets.copy()\n",
    "df_targets_8['drug_name'] = df_drugs['Name'][7]\n",
    "df_targets_8['SMILES'] = df_drugs['SMILES'][7]\n",
    "\n",
    "df_targets_9 = df_targets.copy()\n",
    "df_targets_9['drug_name'] = df_drugs['Name'][8]\n",
    "df_targets_9['SMILES'] = df_drugs['SMILES'][8]\n",
    "\n",
    "targets = pd.concat([df_targets, \n",
    "                  df_targets_2, \n",
    "                  df_targets_3, \n",
    "                  df_targets_4, \n",
    "                  df_targets_5, \n",
    "                  df_targets_6, \n",
    "                  df_targets_7, \n",
    "                  df_targets_8, \n",
    "                  df_targets_9], axis=0)\n",
    "\n",
    "target_name = targets['Protein name'].tolist()\n",
    "target = targets.sequence.tolist()\n",
    "drug_name = targets.drug_name.tolist()\n",
    "drug = targets.SMILES.tolist()\n",
    "\n",
    "targets = pd.read_csv('targets_pre_screen.csv')\n",
    "clean_proteins.to_csv('clean_proteins.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2089543",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from DeepPurpose import DTI as models\n",
    "targets = pd.read_csv('targets_pre_screen.csv')\n",
    "\n",
    "target_name = targets['Protein name'].tolist()\n",
    "target = targets.sequence.tolist()\n",
    "drug_name = targets.drug_name.tolist()\n",
    "drug = targets.SMILES.tolist()\n",
    "\n",
    "# Virtual screening using the trained model or pre-trained model \n",
    "\n",
    "net = models.model_pretrained(model = 'Morgan_AAC_BindingDB_IC50')\n",
    "\n",
    "_ = models.virtual_screening(drug, target, net, drug_name, target_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2c45255b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "fasta = open(\"uniprot_200k.fasta\")\n",
    "df_drugs = pd.read_csv('cann_mols.csv')  \n",
    "\n",
    "fasta_list = list(SimpleFastaParser(fasta))\n",
    "\n",
    "protein_df = pd.DataFrame(fasta_list, columns=['info', 'sequence'])\n",
    "\n",
    "def string_splitter(string):\n",
    "    string = string.split(\"HUMAN\",1)[1]\n",
    "    string = string.split(\"OS=Homo sapiens\")\n",
    "    \n",
    "    return string\n",
    "\n",
    "def info_parser(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info'] # rename the three columns # rename the three columns\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def info_parser_200k(dfx):\n",
    "    df = dfx.copy()\n",
    "    df = df['info'].str.split('|', expand=True) # split on the \"|\" character\n",
    "    df.columns = ['type', 'id', 'info', 'none'] # rename the three columns # rename the three columns\n",
    "    df.drop(\"none\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    return df\n",
    "\n",
    "def info_pre_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of df\n",
    "    list_of_info = list(dfx['info']) # convert column values to list\n",
    "    list_of_info = [string_splitter(x) for x in list_of_info] # apply string_splitter function to list elements\n",
    "    df['temp_col'] = list_of_info # create a temporary column from the processed list\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    split_df = pd.DataFrame(df['temp_col'].tolist(), columns=['Protein name', 'info']) # creating this dataframe to merge back onto processed dataframe\n",
    "    df.drop(\"temp_col\", axis=1, inplace=True) # dropping temp_col\n",
    "    df = pd.concat([df, split_df], axis=1) # merging both dataframes\n",
    "\n",
    "    return df\n",
    "    \n",
    "def info_processed(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.columns = ['type', 'id', 'Protein name', 'drop', 'Species', 'Gene', 'PE', 'Mutation'] # rename columns\n",
    "    df.drop(\"drop\", axis=1, inplace=True) # drop empty column\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df.drop('Species', axis=1, inplace=True) # drop column\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "def info_processed_200k(dfx):\n",
    "    df = dfx.copy() # create a copy of the dataframe\n",
    "    split_df = df['info'].str.split(' ', expand=True) # creating a seperate dataframe to merge with\n",
    "    df = pd.concat([df, split_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "    df = df.drop(df.columns[[0, 3, 4, 6, 7, 8, 9, 10, 11, 12]],axis = 1)\n",
    "\n",
    "    df.columns = ['id', 'Protein name', 'Gene'] # rename columns\n",
    "    df = pd.concat([df, protein_df], axis=1) # merge dataframes\n",
    "    df.drop(\"info\", axis=1, inplace=True) # drop info column\n",
    "\n",
    "    df['Gene'] = df['Gene'].str[3:] # strip first 3 characters\n",
    "    \n",
    "    return df\n",
    "\n",
    "#clean_proteins = info_parser_200k(protein_df)\n",
    "                 #pipe(info_parser).\n",
    "                 #pipe(info_pre_processed).\n",
    "                 #pipe(info_processed))\n",
    "            \n",
    "clean_proteins = (protein_df.\n",
    "                 pipe(info_parser_200k).\n",
    "                 pipe(info_pre_processed).\n",
    "                 pipe(info_processed_200k))\n",
    "\n",
    "targets = clean_proteins[['Protein name','sequence']].copy() # making a copy of the clean dataframe.\n",
    "\n",
    "targets['drug_name'] = df_drugs['Name'][0] # adding a column with a constant value of \"cannaflavin a\".\n",
    "targets['SMILES'] = df_drugs['SMILES'][0] # adding a column with a constant value of \"cannaflavin a\" SMILES string.\n",
    "\n",
    "target_name = targets['Protein name'].tolist()\n",
    "target = targets.sequence.tolist()\n",
    "drug_name = targets.drug_name.tolist()\n",
    "drug = targets.SMILES.tolist()\n",
    "\n",
    "targets.to_csv('targets_pre_screen_cann_a.csv')\n",
    "#clean_proteins.to_csv('clean_proteins.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c9ba4923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein name</th>\n",
       "      <th>sequence</th>\n",
       "      <th>drug_name</th>\n",
       "      <th>SMILES</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T-box 2, isoform CRA_a</td>\n",
       "      <td>MREPALAASAMAYHPFHAPRPADFPMSAFLAAAQPSFFPALALPPG...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Receptor protein-tyrosine kinase</td>\n",
       "      <td>MELQAARACFALLWGCALAAAAAAQGKEVVLLDFAAAGGELGWLTH...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Battenin</td>\n",
       "      <td>MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sirtuin (Silent mating type information regul...</td>\n",
       "      <td>MADEAALALQPGGSPSAAGADREAASSPAGEPLRKRPRRDGPGLER...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Phospholipid phosphatase 1</td>\n",
       "      <td>MFDKTRLPYVALDVLCVLLAGLPFAILTSRHTPFQRGVFCNDESIK...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204901</th>\n",
       "      <td>Probable ATP-dependent RNA helicase DDX5</td>\n",
       "      <td>MSGYSSDRDRGRDRGFGAPRFGGSRAGPLSGKKFGNPGEKLVKKKW...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204902</th>\n",
       "      <td>ELKS/Rab6-interacting/CAST family member 1</td>\n",
       "      <td>MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204903</th>\n",
       "      <td>Voltage-dependent calcium channel subunit alp...</td>\n",
       "      <td>MKLEFLQRKFWAATRQCSTVDGPCTQSCEDSDLDCFVIDNNGFILI...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204904</th>\n",
       "      <td>ELKS/Rab6-interacting/CAST family member 1</td>\n",
       "      <td>MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204905</th>\n",
       "      <td>Mitochondrial enolase superfamily member 1 (F...</td>\n",
       "      <td>XHGADAMHTDPDYSAAYVVIETDAEDGIKGCGITFTLGKGTEVDWS...</td>\n",
       "      <td>Cannflavin A</td>\n",
       "      <td>CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204906 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Protein name  \\\n",
       "0                                 T-box 2, isoform CRA_a    \n",
       "1                       Receptor protein-tyrosine kinase    \n",
       "2                                               Battenin    \n",
       "3        Sirtuin (Silent mating type information regul...   \n",
       "4                             Phospholipid phosphatase 1    \n",
       "...                                                   ...   \n",
       "204901          Probable ATP-dependent RNA helicase DDX5    \n",
       "204902        ELKS/Rab6-interacting/CAST family member 1    \n",
       "204903   Voltage-dependent calcium channel subunit alp...   \n",
       "204904        ELKS/Rab6-interacting/CAST family member 1    \n",
       "204905   Mitochondrial enolase superfamily member 1 (F...   \n",
       "\n",
       "                                                 sequence     drug_name  \\\n",
       "0       MREPALAASAMAYHPFHAPRPADFPMSAFLAAAQPSFFPALALPPG...  Cannflavin A   \n",
       "1       MELQAARACFALLWGCALAAAAAAQGKEVVLLDFAAAGGELGWLTH...  Cannflavin A   \n",
       "2       MGGCAGSRRRFSDSEGEETVPEPRLPLLDHQGAHWKNAVGFWLLGL...  Cannflavin A   \n",
       "3       MADEAALALQPGGSPSAAGADREAASSPAGEPLRKRPRRDGPGLER...  Cannflavin A   \n",
       "4       MFDKTRLPYVALDVLCVLLAGLPFAILTSRHTPFQRGVFCNDESIK...  Cannflavin A   \n",
       "...                                                   ...           ...   \n",
       "204901  MSGYSSDRDRGRDRGFGAPRFGGSRAGPLSGKKFGNPGEKLVKKKW...  Cannflavin A   \n",
       "204902  MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...  Cannflavin A   \n",
       "204903  MKLEFLQRKFWAATRQCSTVDGPCTQSCEDSDLDCFVIDNNGFILI...  Cannflavin A   \n",
       "204904  MYGSARSVGKVEPSSQSPGRSPRLPRSPRLGHRRTNSTGGSSGSSV...  Cannflavin A   \n",
       "204905  XHGADAMHTDPDYSAAYVVIETDAEDGIKGCGITFTLGKGTEVDWS...  Cannflavin A   \n",
       "\n",
       "                                                   SMILES  \n",
       "0       CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "1       CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "2       CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "3       CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "4       CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "...                                                   ...  \n",
       "204901  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "204902  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "204903  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "204904  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "204905  CC(=CCCC(=CCC1=C(C2=C(C=C1O)OC(=CC2=O)C3=CC(=C...  \n",
       "\n",
       "[204906 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
